Engine/ac/string.cpp

```diff
@@ -257,6 +258,55 @@ const char* String_UpperCase(const char *thisString) {
     return CreateNewScriptString(std::move(buf));
 }
 
+void * String_Split(const char *thisString, const char *separator)
+{
+    if ((thisString == nullptr) || (thisString[0] == 0))
+        return nullptr;
+
+    auto const &header = ScriptString::GetHeader((void*)thisString);
+    std::vector<const char*> items{};
+
+    if ((separator == nullptr) || (separator[0] == 0)) {
+        items.push_back(thisString);
+        DynObjectRef arr = DynamicArrayHelpers::CreateStringArray(items);
+        return arr.Obj;
+    }
+
+    size_t seplen = strlen(separator);
+    if (seplen > header.Length) {
+        items.push_back(thisString);
+        DynObjectRef arr = DynamicArrayHelpers::CreateStringArray(items);
+        return arr.Obj;
+    }
+
+    // we will use this list to hold copies of substrings, they will be cleared when this functions exits
+    std::vector<ScriptString::Buffer> buflist = {};
+    const char *ptr = thisString;
+    const char *end = thisString + header.Length;
+    size_t len = 0;
+    while (ptr < end)
+    {
+        const char *found_cstr = strstr(ptr, separator);
+        if (!found_cstr) {
+            found_cstr = end;
+        }
+
+        len = found_cstr - ptr;
+        auto buf = ScriptString::CreateBuffer(len, 0);
+        std::copy(ptr, found_cstr, buf.Get());
+        buf.Get()[len] = 0;
+        buflist.push_back(std::move(buf));
+        items.push_back(buflist.back().Get());
+
+        ptr = found_cstr + seplen;
+    }
+
+    // this takes a vector of const char*, which means null terminated array of strings
+    // null terminated means we have to hold copies instead of references in-string
+    DynObjectRef arr = DynamicArrayHelpers::CreateStringArray(items);
```

Comment:
This DynamicArrayHelpers::CreateStringArray was originally implemented for strings retrieved from somewhere else, but since you already create ScriptString::Buffers here, you can add another variant, which accepts `std::vector<ScriptString> &&` and moves this data to ScriptString::Create().

Reply:
Another alternative would be a variant of DynamicArrayHelpers::CreateArray that accepts a generic std::vector<DynObjectRef> &`. In this case you could create ScriptStrings here directly using ScriptString::Create, and then pass DynObjectRefs into CreateArray, where it makes array, and stores handles in there.

I cannot say which is more convenient at the moment.

Reply:
> since you already create ScriptString::Buffers here, you can add another variant,

I can't figure how to iterate the buffer array from this because I keep getting the following error:

> Call to implicitly-deleted copy constructor of 'ScriptString::Buffer' copy constructor is implicitly deleted because 'Buffer' has a user-declared move constructor

> you could create ScriptStrings here directly using ScriptString::Create, and then pass DynObjectRefs into CreateArray

This worked, so I went this way.

Reply:
Oh so it was double ampersand in the call like `&&items`, that's why it didn't work. :O

Reply:
Sorry, decided to fix that "FIX-ME" and I tried the following:

```
    String type_str = !items.empty() ? items[0].Mgr->GetType() : "";
    auto type = std::find_if(ccInstance::GetRTTI()->GetTypes().begin(), ccInstance::GetRTTI()->GetTypes().end(),
                             [type_str](const RTTI::Type& i) -> bool {return type_str.Equals(i.name); });
```

This works and `type` correctly finds the type "String" in the "_BuiltInScriptHeader.ash", but the type size is 0 - instead of `sizeof(int32_t)`, so I can't use `CreateNew` because I don't understand how to correctly get the size dynamically.

Reply:
The thing is, RTTI may be missing. It has to be decided, whether this API actually requires type info or not.

Reply:
For faster searching for a type there's a lookup table - ccInstance::_rttiLookup, but it may be not exposed right now. May be worth adding an inlined function that returns a constant reference to this table.

Reply:
It's alright, it is working with the CreateOld method, I can update that in a new PR. But thanks for the information!

The find wasn't as slow. About "fully qualified", this means just "_BuiltInScriptHeader.ash::String" or something that can be determined like this right? If so, then it should be fine anyway, at least for internal usage.

Reply:
About this, type's size is 0 because there are no plain fields in it. RTTI only knows the fields declared in script, and there are none in String.

But when you make dynamic array of pointers, you don't need type's size at all, as you don't store type itself in the array, you store pointers to that type. So you ever need only size of pointer. If you were making a dynamic array of primitive types or plain (non-managed) structs, then you would have to get their size.

Reply:
Right, I think this means I need to check the ARRAY_MANAGED_TYPE_FLAG in this case. This flag is found in the type_id here in ags4 and was found in the elementCount in ags3 I think.

I think I understand now how to do it but I think I will leave this for later.